docs/cn/curvefs_volume_space_design.md

```diff
@@ -0,0 +1,535 @@
+
+## 方案对比
+
+### 性能对比
+
+方案一中，Space Server 端负责细粒度的空间分配及持久化，所以当 client 端每次需要申请空间时，都需要发送 RPC 请求。
+
+方案二中，空间分配分为两个层级：粗粒度的空间分配由 MDS 进行，后续细粒度的空间分配在 client 内部处理，无需 RPC 交互。
+
+除 RPC 外，直观来看，持久化、幂等问题处理带来的开销应该是一致的。
+
+### 服务端异常情况下的影响
+
+方案一中，Space Server 在 IO 的核心路径上，所以服务异常会对 IO 产生影响，需要高可用来降低服务异常带来的影响。
+
+方案二中，空间分配的核心流程在 client 端内部处理，只有粗粒度的空间分配需要与 MDS 交互。整体来看，MDS 异常的情况下对 IO 影响较小。
+
+# 结论
+
+空间分配基于方案二进行。
+
+相较于方案一，方案二有以下几点优势:
+
+1. 不需要为空间分配服务重新设计高可用，只需要 MDS 提供高可用的服务即可（目前已实现）。
+2. 空间分配的核心流程在 client 端内部完成，在服务异常的情况下，影响较小。
+3. 空间分配信息直接持久化到卷上。
+
+方案中的遗留问题如下：
+
+1. 空间分配/回收过程中，更新 bitmap 和更新 inode 不是原子操作，再此过程中进程异常退出，会导致文件系统不一致。
+2. fuse-client 申请 block group 时可能会出现已完全分配的情况。
+
+## 第一阶段开发计划
+
+1. 完成基本的空间分配、回收功能
+2. 遗留问题先进行规避
+   1. 空间分配过程中进程异常退出后，先更新 bitmap，后更新 inode 会造成空间浪费的问题。**第一阶段先不处理**。
+   2. 空间回收过程中进程异常退出后，通过先更新 inode，再清理 bitmap 的方式，规避反向操作可能带来的不一致问题，造成的空间浪费问题，**第一阶段先不处理**。按照这种方式处理的时序图如下：
+![curvefs space block group free](../images/curvefs-space-block-group-free.png)
+
+   3. fuse-client 申请 block group 时，如果 block group 完全分配，进行主动的 extend 卷。
+   
+   
+### 相关 RPC 接口
+
+```cpp
+syntax="proto2";
+package curvefs.mds.space;
+option cc_generic_services=true;
+
+enum SpaceStatusCode {
+    SpaceOk = 0;
+    SpaceUnknownError = 1;
+    SpaceNoSpace = 2;
+    SpaceNotFound = 3;
+}
+
+enum BitmapLocation {
+    AtStart = 1;
+    AtEnd = 2;
+}
+
+message BlockGroup {
+    // start offset on volume
+    required uint64 offset = 1;
+    // total bytes of each block group
+    // this must be fixed when create fs or mkfs, so, this field maybe not needed
+    required uint32 size = 2;
+    // used bytes
+    required uint32 used = 3;
+    // available bytes
+    required uint32 available = 4;
+    // block size
+    // this must be fixed when create fs or mkfs, so, this field maybe not needed
+    required uint32 blocksize = 5;
+    // bitmap locaiton
+    required BitmapLocation bitmap = 6;
+    // owner, who owns this block group
+    optional string owner = 7;
+}
+
+message AcquireBlockGroupRequest {
+    required uint32 fsId = 1;
+    required string owner = 2;    // or just ip:port
+    required uint64 size = 3;     // or number of block groups
+}
+
+message AcquireBlockGroupResponse {
+    required SpaceStatusCode status = 1;
+    optional uint64 size = 2;
+    repeated BlockGroup blockGroups = 3;
+}
+
+message ReleaseBlockGroupRequest {
+    required uint32 fsId = 1;
+    repeated BlockGroup blockGroups = 2;
+}
+
+message ReleaseBlockGroupResponse {
+    required SpaceStatusCode status = 1;
+}
+
+message StatSpaceRequest {
+    required uint32 fsId = 1;
+}
+
+message SpaceInfo {
+    required uint64 size = 1;
+    required uint64 used = 2;
+    required uint64 avilable = 3;
+}
+
+message StatSpaceResponse {
+    required SpaceStatusCode status = 1;
+    optional SpaceInfo spaceInfo = 2;
+}
+
+message UpdateUsageRequest {
+    required uint32 fsId = 1;
+    repeated BlockGroup blockGroups = 2;
```

Comment:
Do you replace the existing blockgroup with a new one, or update some of its fields?